lib/MusicBrainz/Server/Data/EntityTag.pm

```diff
@@ -108,13 +107,71 @@ sub find_user_tags_for_entities
     return @tags;
 }
 
+sub find_genres_for_entities
+{
+    my ($self, @ids) = @_;
+
+    return unless scalar @ids;
+
+    my $query = "SELECT tag.id as tag_id, tag.name, entity_tag.count,
+                        entity_tag.".$self->type." AS entity, genre.id as genre_id
+                 FROM " . $self->tag_table . " entity_tag
+                 JOIN tag ON tag.id = entity_tag.tag
+                 JOIN genre ON tag.name = genre.name
+                 WHERE " . $self->type . " IN (" . placeholders(@ids) . ")
+                 ORDER BY musicbrainz_collate(tag.name)";
+
+    my @tags = $self->query_to_list($query, \@ids);
+
+    $self->c->model('Genre')->load(map { $_->tag } @tags);
+
+    return @tags;
+}
+
+sub find_user_genres_for_entities
+{
+    my ($self, $user_id, @ids) = @_;
+
+    return unless scalar @ids;
+
+    my $type = $self->type;
+    my $table = $self->tag_table . '_raw';
+    my $query = "SELECT entity_tag.tag as tag_id, $type AS entity,
+                        is_upvote
+                 FROM $table entity_tag
+                 JOIN tag ON tag.id = entity_tag.tag
+                 JOIN genre ON tag.name = genre.name
+                 WHERE editor = ?
+                 AND $type IN (" . placeholders(@ids) . ")
+                 ORDER BY musicbrainz_collate(tag.name)";
+
+    my @tags = $self->query_to_list($query, [$user_id, @ids], sub {
+        my ($model, $row) = @_;
+        return MusicBrainz::Server::Entity::UserTag->new(
+            tag_id    => $row->{tag_id},
+            editor_id => $user_id,
+            entity_id => $row->{entity},
+            is_upvote => $row->{is_upvote},
+        );
+    });
+
+    $self->c->model('Tag')->load(@tags);
```

Comment:
You actually have all the info needed to construct each tag in the query, so could just add a `tag => MusicBrainz::Server::Entity::Tag->new(...)` key/value pair to the `UserTag` construction above.

Reply:
Did that as a separate commit.